File: 2L-HFVRP/src/Fekete/lbFekete.java
package Fekete;
/*
 * From Sergey
 */
import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.ArrayList;
import objects.Customer;
import objects.Item;
import objects.VehicleType;

public class lbFekete
{

	public static void scaleGeneration(BigDecimal start, BigDecimal step,
			ArrayList<Customer> customers, ArrayList<VehicleType> vehicleTypes)
	{
		int ptr = -1;
		for (VehicleType type : vehicleTypes)
		{
			ptr++;
			for (Customer customer : customers)
			{
				customer.scales.add(new ArrayList<Double>());
				if (!customer.isSuitableForType(ptr))
					continue;
				BigDecimal stop = new BigDecimal("0.5");
				double value;
				for (BigDecimal r = start; r.compareTo(stop) <= 0; r = r.add(step))
				{
					value = 0;
					for (Item item : customer.get_items())
						value += scale1(item.get_length(), item.get_width(), 1, r,
								type.get_length(), type.get_width()).doubleValue();
					customer.scales.get(ptr).add(value);

					value = 0;
					for (Item item : customer.get_items())
						value += scale2(item.get_length(), item.get_width(), 1, r,
								type.get_length(), type.get_width()).doubleValue();
					customer.scales.get(ptr).add(value);

					value = 0;
					for (Item item : customer.get_items())
						value += scale3(item.get_length(), item.get_width(), 1, r,
								type.get_length(), type.get_width()).doubleValue();
					customer.scales.get(ptr).add(value);

					value = 0;
					for (Item item : customer.get_items())
						value += scale4(item.get_length(), item.get_width(), 1, r,
								type.get_length(), type.get_width()).doubleValue();
					customer.scales.get(ptr).add(value);

					value = 0;
					for (Item item : customer.get_items())
						value += scale5(item.get_length(), item.get_width(), r, type.get_length(),
								type.get_width()).doubleValue();
					customer.scales.get(ptr).add(value);

					value = 0;
					for (Item item : customer.get_items())
						value += scale6(item.get_length(), item.get_width(), r, type.get_length(),
								type.get_width()).doubleValue();
					customer.scales.get(ptr).add(value);

					for (BigDecimal q = start; q.compareTo(stop) <= 0; q = q.add(step))
					{
						value = 0;
						for (Item item : customer.get_items())
							value += scale7(item.get_length(), item.get_width(), r, q,
									type.get_length(), type.get_width()).doubleValue();
						customer.scales.get(ptr).add(value);
					}
				}
			}
		}
	}

	private static BigDecimal scale1(int h, int w, int k, BigDecimal r, int L, int W)
	{
		return uFunction(new BigDecimal(h).divide(new BigDecimal(L), 8, BigDecimal.ROUND_HALF_UP))
				.multiply(
						UFunction(new BigDecimal(w).divide(new BigDecimal(W), 8,
								BigDecimal.ROUND_HALF_UP), r));
	}

	private static BigDecimal scale2(int h, int w, int k, BigDecimal r, int L, int W)
	{
		return uFunction(new BigDecimal(w).divide(new BigDecimal(W), 8, BigDecimal.ROUND_HALF_UP))
				.multiply(
						UFunction(new BigDecimal(h).divide(new BigDecimal(L), 8,
								BigDecimal.ROUND_HALF_UP), r));
	}

	private static BigDecimal scale3(int h, int w, int k, BigDecimal r, int L, int W)
	{
		return uFunction(new BigDecimal(h).divide(new BigDecimal(L), 8, BigDecimal.ROUND_HALF_UP))
				.multiply(
						fFunction(new BigDecimal(w).divide(new BigDecimal(W), 8,
								BigDecimal.ROUND_HALF_UP), r));
	}

	private static BigDecimal scale4(int h, int w, int k, BigDecimal r, int L, int W)
	{
		return uFunction(new BigDecimal(w).divide(new BigDecimal(W), 8, BigDecimal.ROUND_HALF_UP))
				.multiply(
						fFunction(new BigDecimal(h).divide(new BigDecimal(L), 8,
								BigDecimal.ROUND_HALF_UP), r));
	}

	private static BigDecimal scale5(int h, int w, BigDecimal r, int L, int W)
	{
		return (new BigDecimal(h).divide(new BigDecimal(L), 8, BigDecimal.ROUND_HALF_UP))
				.multiply(UFunction(
						new BigDecimal(w).divide(new BigDecimal(W), 8, BigDecimal.ROUND_HALF_UP), r));
	}

	private static BigDecimal scale6(int h, int w, BigDecimal r, int L, int W)
	{
		return (new BigDecimal(w).divide(new BigDecimal(W), 8, BigDecimal.ROUND_HALF_UP))
				.multiply(UFunction(
						new BigDecimal(h).divide(new BigDecimal(L), 8, BigDecimal.ROUND_HALF_UP), r));
	}

	private static BigDecimal scale7(int h, int w, BigDecimal r, BigDecimal q, int L, int W)
	{
		return fFunction(new BigDecimal(h).divide(new BigDecimal(L), 8, BigDecimal.ROUND_HALF_UP),
				r).multiply(
				fFunction(new BigDecimal(w).divide(new BigDecimal(W), 8, BigDecimal.ROUND_HALF_UP),
						q));
	}

	private static BigDecimal uFunction(BigDecimal xVar)
	{
		if (frac(xVar.multiply(new BigDecimal(2))).compareTo(new BigDecimal(0)) == 0)
			return xVar;
		else
			return xVar.multiply(new BigDecimal(2)).setScale(0, RoundingMode.FLOOR);
	}

	private static BigDecimal UFunction(BigDecimal xVar, BigDecimal param)
	{
		if (xVar.compareTo((new BigDecimal(1)).subtract(param)) > 0)
			return (new BigDecimal(1));
		if ((param.compareTo(xVar) <= 0)
				&& (xVar.compareTo((new BigDecimal(1)).subtract(param)) <= 0))
			return xVar;
		if (xVar.compareTo(param) < 0)
			return (new BigDecimal(0));
		return (new BigDecimal(0));
	}

	private static BigDecimal fFunction(BigDecimal xVar, BigDecimal param)
	{
		if (xVar.compareTo(new BigDecimal("0.5")) > 0)
			return (new BigDecimal(1)).subtract(((new BigDecimal(1)).subtract(xVar).divide(param,
					8, BigDecimal.ROUND_HALF_UP)).setScale(0, RoundingMode.FLOOR).divide(
					(new BigDecimal(1)).divide(param, 8, BigDecimal.ROUND_HALF_UP).setScale(0,
							RoundingMode.FLOOR), 8, BigDecimal.ROUND_HALF_UP));
		if ((param.compareTo(xVar) <= 0) && (xVar.compareTo(new BigDecimal("0.5")) <= 0))
			return (new BigDecimal(1)).divide(
					(new BigDecimal(1)).divide(param, 8, BigDecimal.ROUND_HALF_UP).setScale(0,
							RoundingMode.FLOOR), 8, BigDecimal.ROUND_HALF_UP);
		if (xVar.compareTo(param) < 0)
			return (new BigDecimal(0));
		return (new BigDecimal(0));
	}

	private static BigDecimal frac(BigDecimal xVar)
	{
		return xVar.subtract(xVar.setScale(0, RoundingMode.FLOOR));
	}
}

File: VRPC#/ConsoleApplication1/ConsoleApplication1/Fekete/lbFekete.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ConsoleApplication1.Fekete
{
   
    
        /*
         * From Sergey
         */
        using Customer = objects.Customer;
        using Item = objects.Item;
        using VehicleType = objects.VehicleType;

        public class lbFekete
        {

            public static void scaleGeneration(decimal start, decimal step, List<Customer> customers, List<VehicleType> vehicleTypes)
            {
                int ptr = -1;
                foreach (VehicleType type in vehicleTypes)
                {
                    ptr++;
                    foreach (Customer customer in customers)
                    {
                        customer.scales.Add(new List<double?>());
                        if (!customer.isSuitableForType(ptr))
                        {
                            continue;
                        }
                        decimal stop = new decimal("0.5");
                        double value;
                        for (decimal r = start; r.CompareTo(stop) <= 0; r = r + step)
                        {
                            value = 0;
                            foreach (Item item in customer.get_items())
                            {
                                value += (double)scale1(item.get_length(), item.get_width(), 1, r, type.get_length(), type.get_width());
                            }
                            customer.scales[ptr].Add(value);

                            value = 0;
                            foreach (Item item in customer.get_items())
                            {
                                value += (double)scale2(item.get_length(), item.get_width(), 1, r, type.get_length(), type.get_width());
                            }
                            customer.scales[ptr].Add(value);

                            value = 0;
                            foreach (Item item in customer.get_items())
                            {
                                value += (double)scale3(item.get_length(), item.get_width(), 1, r, type.get_length(), type.get_width());
                            }
                            customer.scales[ptr].Add(value);

                            value = 0;
                            foreach (Item item in customer.get_items())
                            {
                                value += (double)scale4(item.get_length(), item.get_width(), 1, r, type.get_length(), type.get_width());
                            }
                            customer.scales[ptr].Add(value);

                            value = 0;
                            foreach (Item item in customer.get_items())
                            {
                                value += (double)scale5(item.get_length(), item.get_width(), r, type.get_length(), type.get_width());
                            }
                            customer.scales[ptr].Add(value);

                            value = 0;
                            foreach (Item item in customer.get_items())
                            {
                                value += (double)scale6(item.get_length(), item.get_width(), r, type.get_length(), type.get_width());
                            }
                            customer.scales[ptr].Add(value);

                            for (decimal q = start; q.CompareTo(stop) <= 0; q = q + step)
                            {
                                value = 0;
                                foreach (Item item in customer.get_items())
                                {
                                    value += (double)scale7(item.get_length(), item.get_width(), r, q, type.get_length(), type.get_width());
                                }
                                customer.scales[ptr].Add(value);
                            }
                        }
                    }
                }
            }

            private static decimal scale1(int h, int w, int k, decimal r, int L, int W)
            {
                return uFunction((new decimal(h)).divide(new decimal(L), 8, decimal.ROUND_HALF_UP)) * UFunction((new decimal(w)).divide(new decimal(W), 8, decimal.ROUND_HALF_UP), r);
            }

            private static decimal scale2(int h, int w, int k, decimal r, int L, int W)
            {
                return uFunction((new decimal(w)).divide(new decimal(W), 8, decimal.ROUND_HALF_UP)) * UFunction((new decimal(h)).divide(new decimal(L), 8, decimal.ROUND_HALF_UP), r);
            }

            private static decimal scale3(int h, int w, int k, decimal r, int L, int W)
            {
                return uFunction((new decimal(h)).divide(new decimal(L), 8, decimal.ROUND_HALF_UP)) * fFunction((new decimal(w)).divide(new decimal(W), 8, decimal.ROUND_HALF_UP), r);
            }

            private static decimal scale4(int h, int w, int k, decimal r, int L, int W)
            {
                return uFunction((new decimal(w)).divide(new decimal(W), 8, decimal.ROUND_HALF_UP)) * fFunction((new decimal(h)).divide(new decimal(L), 8, decimal.ROUND_HALF_UP), r);
            }

            private static decimal scale5(int h, int w, decimal r, int L, int W)
            {
                return ((new decimal(h)).divide(new decimal(L), 8, decimal.ROUND_HALF_UP)).multiply(UFunction((new decimal(w)).divide(new decimal(W), 8, decimal.ROUND_HALF_UP), r));
            }

            private static decimal scale6(int h, int w, decimal r, int L, int W)
            {
                return ((new decimal(w)).divide(new decimal(W), 8, decimal.ROUND_HALF_UP)).multiply(UFunction((new decimal(h)).divide(new decimal(L), 8, decimal.ROUND_HALF_UP), r));
            }

            private static decimal scale7(int h, int w, decimal r, decimal q, int L, int W)
            {
                return fFunction((new decimal(h)).divide(new decimal(L), 8, decimal.ROUND_HALF_UP), r) * fFunction((new decimal(w)).divide(new decimal(W), 8, decimal.ROUND_HALF_UP), q);
            }

            private static decimal uFunction(decimal xVar)
            {
                if (frac(xVar * (new decimal(2))).CompareTo(new decimal(0)) == 0)
                {
                    return xVar;
                }
                else
                {
                    return xVar * (new decimal(2)).setScale(0, RoundingMode.FLOOR);
                }
            }

            private static decimal UFunction(decimal xVar, decimal param)
            {
                if (xVar.CompareTo((new decimal(1)) - param) > 0)
                {
                    return (new decimal(1));
                }
                if ((param.CompareTo(xVar) <= 0) && (xVar.CompareTo((new decimal(1)) - param) <= 0))
                {
                    return xVar;
                }
                if (xVar.CompareTo(param) < 0)
                {
                    return (new decimal(0));
                }
                return (new decimal(0));
            }

            private static decimal fFunction(decimal xVar, decimal param)
            {
                if (xVar.CompareTo(new decimal("0.5")) > 0)
                {
                    return (new decimal(1)) - (((new decimal(1)) - xVar.divide(param, 8, decimal.ROUND_HALF_UP)).setScale(0, RoundingMode.FLOOR).divide((new decimal(1)).divide(param, 8, decimal.ROUND_HALF_UP).setScale(0, RoundingMode.FLOOR), 8, decimal.ROUND_HALF_UP));
                }
                if ((param.CompareTo(xVar) <= 0) && (xVar.CompareTo(new decimal("0.5")) <= 0))
                {
                    return (new decimal(1)).divide((new decimal(1)).divide(param, 8, decimal.ROUND_HALF_UP).setScale(0, RoundingMode.FLOOR), 8, decimal.ROUND_HALF_UP);
                }
                if (xVar.CompareTo(param) < 0)
                {
                    return (new decimal(0));
                }
                return (new decimal(0));
            }

            private static decimal frac(decimal xVar)
            {
                return xVar - xVar.setScale(0, RoundingMode.FLOOR);
            }
        }
    }